### PR TITLE
Improved DDP checkpoint documentation

### LINKED ISSUE #84589: Update DDP checkpoint documentation
### 📚 The doc issue

DDP checkpoint documentation here: https://github.com/pytorch/pytorch/blob/88b1cc885cc92b9483eec95546bb48c7bccea070/torch/nn/parallel/distributed.py#L333 should be updated in light of new non-reentrant checkpoint API available. 

In particular, checkpoint with non_reentrant=True supports the use cases that are mentioned as unsupported.

### Suggest a potential alternative/fix

_No response_

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501

### PR DESCRIPTION
**Pull request by andreigh**
Amended the documentation for the specified case.

Fixes #84589

### CODE REVIEW

**Comment by awgu** [comment]
Are the CMakeLists.txt changes intended?

**Comment by andreigh** [comment]
> Are the CMakeLists.txt changes intended?

Wrong push - fixed now

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -456,12 +456,15 @@ class DistributedDataParallel(Module, Joinable):
...
-        checkpointing with :meth:`torch.utils.checkpoint`. DDP will work as
-        expected when there are no unused parameters in the model and each layer
+        checkpointing with :meth:`torch.utils.checkpoint`.
+        If the checkpoint is done with use_reentrant=True (the default), DDP will work
+        as expected when there are no unused parameters in the model and each layer
         is checkpointed at most once (make sure you are not passing
         `find_unused_parameters=True` to DDP). We currently do not support the
         case where a layer is checkpointed multiple times, or when there unused
         parameters in the checkpointed model.
+        If, however, the checkpoint is done with use_reentrant=False, DDP will work
```
Comment (wanchaol) [inline torch/nn/parallel/distributed.py:466]: I think we should change our documentation to officially recommend `use_reentrant=False` as that's the recommended way of `torch.utils.checkpoint`, and discourage `use_reentrant=True` usage. cc @rohan-varma

Comment (andreigh) [reply torch/nn/parallel/distributed.py]: I've updated the documentation

**Review by wanchaol — APPROVED** [review]

**Review by fduwjj — APPROVED** [review]
Please make sure CI all passed. Thanks!

**Comment by andreigh** [comment]
@pytorch rebase

**Comment by andreigh** [comment]
@fduwjj CI is green !

**Comment by kit1980** [comment]
> CI is green !

@andreigh FYI after a PR approved by a project member, anyone can merge the PR by posting "@pytorchbot merge" comment.

**Pull request closed without merging**